movai_core_shared/core/zmq_server.py

```diff
@@ -0,0 +1,166 @@
+        if not isinstance(debug, bool):
+            raise ValueError("debug must be of type bool.")
+        self._name = server_name
+        self._addr = bind_addr
+        self._logger = logging.getLogger(server_name)
+        self._new_loop = new_loop
+        self._debug = debug
+        self._initialized = False
+        self._running = False
+        self._ctx = None
+        self._socket = None
+
+    def _set_context(self) -> None:
+        """Initializes the zmq context."""
+        self._ctx = zmq.asyncio.Context()
+        self._socket = self._ctx.socket(zmq.ROUTER)
+        if self._socket is None:
+            raise zmq.ZMQError(msg= "Failed to create socket")
+
+    def _bind(self) -> None:
+        """Binds the zmq socket to the appropriate file/address."""
+        self._socket.bind(self._addr)
+        self._logger.info(f"{self.__class__.__name__} is listening on {self._addr}")
+
+    async def _accept(self) -> None:
+        """accepts new connections requests to zmq."""
+        while self._running:
+            try:
+                if self._debug:
+                    self._logger.debug("Waiting for new requests.\n")
+                request = await self._socket.recv_multipart()
+                asyncio.create_task(self._handle(request))
+            except Exception as error:
+                self._logger.error(f"ZMQServer Error: {str(error)}")
+                continue
+
+    async def _handle(self, msg_buffer) -> None:
+        index = len(msg_buffer) - 1
+        buffer = msg_buffer[index]
+
+        if buffer is None:
+            self._logger.warning("Request has no buffer, can't handle request.")
+            return
+
+        request_msg = json.loads(buffer)
+        if "request" not in request_msg:
+            raise UnknownRequestError(f"The message format is unknown: {request_msg}.")
+        request = request_msg.get("request")
+        
+        if self._debug:
+            general_request = Request(**request)
+            self._logger.debug(general_request.__str__())
+        
+        response = await self.handle_request(request)
+        response_required = request.get("response_required")
+        if response_required:
+            if response.get("response") is None:
+                response = {"response": response}
+            response = await self.handle_response(response)
+            index = len(msg_buffer) - 1
+            msg_buffer[index] = json.dumps(response).encode("utf8")
+            self._socket.send_multipart(msg_buffer)
+            if self._debug:
+                self._logger.debug(f"{self._name} successfully sent a respone.")
+                
+
+
+    def _close(self) -> None:
+        """close the zmq socket."""
+        self._socket.close()
+        self._ctx.destroy()
+        self._socket = None
+        self._ctx = None
+        self._initialized = False
+
+    def __del__(self):
+        """closes the socket when the object is destroyed.
+        """
+        self._close()
```

Comment:
why separte to different methods
are you gonna use _close without deleting the object?

Reply:
in case of error I might want to stop the server and make sure all zmq resources got released, and later I may try run the object again..

Reply:
so why is that private
(_ before)